include/deal.II/matrix_free/tools.h

```diff
@@ -962,9 +1284,52 @@ namespace MatrixFreeTools
         locally_relevant_constraints_hn_map;
 
       // scratch array
+      AlignedVector<VectorizedArrayType> temp_values;
       AlignedVector<VectorizedArrayType> values_dofs;
     };
 
+    template <bool is_face,
+              int  dim,
+              typename Number,
+              typename VectorizedArrayType>
+    bool
+    is_fe_nothing(
+      const MatrixFree<dim, Number, VectorizedArrayType> &matrix_free,
+      const std::pair<unsigned int, unsigned int>        &range,
+      const unsigned int                                  dof_no,
+      const unsigned int                                  quad_no,
+      const unsigned int first_selected_component,
+      const unsigned int fe_degree,
+      const unsigned int n_q_points_1d,
+      const bool         is_interior_face = true)
+    {
+      const unsigned int static_n_q_points =
+        is_face ? Utilities::pow(n_q_points_1d, dim - 1) :
+                  Utilities::pow(n_q_points_1d, dim);
+
+      unsigned int active_fe_index = 0;
+      if (!is_face)
+        active_fe_index = matrix_free.get_cell_active_fe_index(range);
+      else if (is_interior_face)
+        active_fe_index = matrix_free.get_face_range_category(range).first;
+      else
+        active_fe_index = matrix_free.get_face_range_category(range).second;
+
+      const auto init_data = dealii::internal::
+        extract_initialization_data<is_face, dim, Number, VectorizedArrayType>(
+          matrix_free,
+          dof_no,
+          first_selected_component,
+          quad_no,
+          fe_degree,
+          static_n_q_points,
+          active_fe_index,
+          numbers::invalid_unsigned_int /*active_quad_index*/,
+          numbers::invalid_unsigned_int /*face_type*/);
```

Comment:
Is this the simplest way to get the case of `FE_Nothing`? I would have looked up the active fe index and then checked the `matrix_free.get_dof_handler(dof_no).get_fe_collection()[active_fe_index]`? Is there something not working, e.g. for the faces?

Reply:
The reason why we do this is that we did not copy the (somewhat complicated) logic in:

https://github.com/dealii/dealii/blob/f8df7d6869db0d83ded64008c9adeae170b52b5b/include/deal.II/matrix_free/fe_evaluation.h#L2291-L2297

If we would copy these lines, I am afraid that the codes will diverge sometime.  @kronbichler  What is your opinion on this?

Reply:
So in other words, the `active_fe_index` variable you have at your disposal here is not the index we need to check within the actual `FECollection` class in the `DoFHandler`, right? In that case, I think I agree with the code here, even though it is not pretty.